File: src/game/slot/netent/flowers/flowers_link.ts
import {
  AlgInfo,
  GameType,
  GameProps,
  GameAlias,
  AlgDescr,
  makeRtpList,
  registerGame,
  GameRegistration,
  Gamble
} from '../../../types'; // Main game types
import { NewFlowersGame } from './flowers_rule';
import { calcStatReg } from './flowers_calc';
import { ReelsMap } from './flowers_reels';

const FlowersAliases: GameAlias[] = [
  { prov: "NetEnt", name: "Flowers", date: "2013-11-11" },
];

const FlowersAlgDescr: AlgDescr = {
  gt: GameType.GTslot,
  gp: GameProps.GPlpay |
      GameProps.GPlsel |
      GameProps.GPretrig |
      GameProps.GPfgreel |
      GameProps.GPfgmult |
      GameProps.GPwild,
  sx: 5, // Screen X dimension (reels)
  sy: 3, // Screen Y dimension (rows)
  sn: 17, // Number of symbols in LinePay (seems to be max symbol ID from flowers_rule.go LinePay array size)
  ln: 30, // Number of bet lines (from BetLines.length in flowers_rule.go)
  bn: 0,  // Number of bonus games (0 for Flowers, as per Go)
  rtp: makeRtpList(ReelsMap), // Create RTP list from ReelsMap
};

export const FlowersGameInfo: AlgInfo = {
  aliases: FlowersAliases,
  algDescr: FlowersAlgDescr,
  // gameFactory and calcStat will be part of the registration
};

// Equivalent to Go's init() function for game registration
function registerFlowersGame(): void {
  const registration: GameRegistration = {
    info: FlowersGameInfo,
    factory: NewFlowersGame as () => Gamble, // Ensure NewFlowersGame matches Gamble type
    statCalc: (ctx: any, mrtp: number) => calcStatReg(ctx, mrtp) // Wrap for context if needed
  };
  registerGame("NetEnt/Flowers", registration);
}

// Call the registration function to make the game available
registerFlowersGame();

// Optional: export a function to get the info, or rely on the registration system
export function getFlowersGameInfo(): AlgInfo {
  return FlowersGameInfo;
}

File: game/slot/netent/flowers/flowers_link.go
//go:build !prod || full || netent

package flowers

import (
	_ "embed"

	"github.com/slotopol/server/game"
)

//go:embed flowers_data.yaml
var data []byte

var Info = game.AlgInfo{
	Aliases: []game.GameAlias{
		{Prov: "NetEnt", Name: "Flowers", LNum: 30, Date: game.Date(2013, 11, 11)}, // see: https://games.netent.com/video-slots/flowers/
	},
	AlgDescr: game.AlgDescr{
		GT: game.GTslot,
		GP: game.GPlpay |
			game.GPlsel |
			game.GPfgseq |
			game.GPfgreel |
			game.GPfgmult |
			game.GPwild,
		SX: 5,
		SY: 3,
		SN: sn,
		LN: len(BetLines),
		BN: 0,
	},
	Update: func(ai *game.AlgInfo) { ai.RTP = game.MakeRtpList(ReelsMap) },
}

func init() {
	Info.SetupFactory(func(sel int) game.Gamble { return NewGame(sel) }, CalcStatReg)
	game.DataRouter["netent/flowers/bon"] = &ReelsBon
	game.DataRouter["netent/flowers/reel"] = &ReelsMap
	game.LoadMap = append(game.LoadMap, data)
}
